### DISCUSSION TITLE
Set scaleFactor to UI scale setting for modded effect UIs.

### DISCUSSION THREAD

**Discussion by DanielWillett** [SmartlyDressedGames/Unturned-3.x-Community#3967] [Ideas]
Automatically change the `Canvas Scaler`'s scale based on the user's UI scale setting.

All this would need a few lines in createUIEffect and a new asset field:

```cs
// make this opt-in from the asset so it doesn't break old UIs.
if (effectAsset.autoScale)
{
   CanvasScaler? scaler = transform.GetComponentInChildren<CanvasScaler>();

   if (scaler != null && scaler.uiScaleMode == CanvasScaler.ScaleMode.ConstantPixelSize)
   {
        scaler.scaleFactor = GraphicsSettings.userInterfaceScale;
   }
}
```

**Comment by liebesleid-eth** [comment]
Please, this would be amazing π•π

**Comment by negrifelipe** [comment]
Good too imo

**Comment by DanielWillett** [comment]
@SDGNelson since you're doing UI changes next update any chance you could reconsider adding this?

It makes it possible for us to line up UI elements in custom UI's with vanilla UI. Currently, there's not really a catch-all way to do this.

**Comment by DanielWillett** [comment]
:(

**Comment by Corbyncc** [comment]
Would be nice for some feedback on this @SDGNelson

**Comment by DanielWillett** [comment]
@SDGNelson sorry to bug you so much here but this would really improve modded UIs in the mean time, I know it seems small but right now there's no way to 'snap' UIs to vanilla UIs (status icons, etc.) because of scaling, and the automated scaling modes in Unity kinda suck compared to this.

This is like a 5 minute change with no public API changes and would drastically improve the quality and add some more possibilities with modded UI. There's really no reason it has to update automatically when the setting is changed either.

You really don't even need to add an asset field since as far as I know no one actually uses ConstantPixelSize for scaling anyway.

Would at least be nice to know why you don't want to add it.

\- thanks

**Reply by SDGNelson** [reply]
ok ok ok I'll change it to add a component that keeps the canvas scaler in sync with the game's UI scale setting for the next update.

**Reply by DanielWillett** [reply]
much appreciated

**Reply by SDGNelson** [reply]
I do fully agree plugin UIs deserve better support BTW. It's just a messy situation all-around.
